skrub/_dispatch.py

```diff
@@ -0,0 +1,194 @@
+
+
+>>> def my_other_function(df):
+...     # not ok: will fail if df is a polars DataFrame
+...     df.values
+...     # not ok: only happens to work by chance because both polars and pandas use
+...     # the same name with the same signature and semantics, but could fail for
+...     # other backends added in the future
+...     df.shape
+...     # ok
+...     import skrub._dataframe as ns
+...     ns.shape(df)
+
+
+We can inspect which implementations were registered in ``my_function.registry``.
+
+It is also possible to register a specialization specifically for dataframes or
+for columns:
+
+>>> @my_function.specialize("pandas", "Column")
+... def _my_function_pandas_column(df, other_arg=1.0):
+...     print("calling pandas Series implementation")
+...     print("               ^^^^^^               ")
+
+>>> my_function(pd.Series([0]))
+calling pandas Series implementation
+               ^^^^^^
+>>> my_function(pd.DataFrame())
+calling pandas implementation
+
+Note that the last registered implementation wins -- there is no magic to
+prioritize specializations based on how specific they are.
+
+>>> @my_function.specialize("pandas")
+... def _override_all_pandas_specializations(df, other_arg=1.0):
+...     print("Any pandas (Series or DataFrame)")
+
+>>> my_function(pd.Series([0]))
+Any pandas (Series or DataFrame)
+>>> my_function(pd.DataFrame())
+Any pandas (Series or DataFrame)
+
+Importantly, single dispatch is done based on the type of the
+**first argument**. This means our generic function should be written in such a way
+that the first argument is the dataframe or column that will determine which
+implementation to use. For example the following would not work:
+
+>>> @dispatch
+... def sample(n_rows, df):
+...     ...
+
+Instead we must rewrite it to be:
+
+>>> @dispatch
+... def sample(df, n_rows):
+...     ...
+"""
+
+from functools import singledispatch
+
+
+def _load_dataframe_module_info(name):
+    # if the module is not installed, import errors are propagated
+    if name == "pandas":
+        import pandas
+
+        return {
+            "module": pandas,
+            "types": {
+                "DataFrame": [pandas.DataFrame],
+                "Column": [pandas.Series],
+            },
+        }
+    if name == "polars":
+        import polars
+
+        return {
+            "module": polars,
+            "types": {
+                "DataFrame": [polars.DataFrame, polars.LazyFrame],
+                "LazyFrame": [polars.LazyFrame],
+                "EagerFrame": [polars.DataFrame],
+                "Column": [polars.Series],
+            },
+        }
+    raise KeyError(f"Unknown module: {name}")
+
+
+def dispatch(function):
+    dispatched = singledispatch(function)
+
+    def specialize(module_name, generic_type_names=None):
+        try:
+            module_info = _load_dataframe_module_info(module_name)
+        except (ImportError, KeyError):
+
+            def decorator(specialized_impl):
```

Comment:
I think this worth a comment here. If I'm not wrong this is falling back the generic implementation but we call this `specialized_impl` even in this case.

Reply:
Here `specialized_impl` is the specialized implementation , eg `_drop_nulls_polars`. If polars isn't installed, we just return the function without registering it: we have no way of registering it and it would never get called anyway. so `_drop_nulls_polars` will exist in the module, but it will not appear in the generic function's registry, and it will never be called. If polars _is_ installed, we register it first but we also return the unmodified function afterwards